orchagent/tunneldecaporch.cpp

```diff
@@ -249,44 +287,420 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
             if (valid && !exists)
             {
 
-                if (addDecapTunnel(key, tunnel_type, ip_addresses, p_src_ip, dscp_mode, ecn_mode, encap_ecn_mode, ttl_mode,
-                                    dscp_to_tc_map_id, tc_to_pg_map_id))
+                if (addDecapTunnel(key, tunnel_type, p_src_ip, dscp_mode, ecn_mode, encap_ecn_mode, ttl_mode,
+                                   dscp_to_tc_map_id, tc_to_pg_map_id))
                 {
                     // Record only
                     tunnelTable[key].encap_tc_to_dscp_map_id = tc_to_dscp_map_id;
                     tunnelTable[key].encap_tc_to_queue_map_id = tc_to_queue_map_id;
-                    SWSS_LOG_NOTICE("Tunnel(s) added to ASIC_DB.");
+                    SWSS_LOG_NOTICE("Tunnel %s added to ASIC_DB.", key.c_str());
+
+                    // process unhandled decap terms
+                    processUnhandledDecapTunnelTerms(key);
                 }
                 else
                 {
-                    SWSS_LOG_ERROR("Failed to add tunnels to ASIC_DB.");
+                    SWSS_LOG_ERROR("Failed to add tunnel %s to ASIC_DB.", key.c_str());
                 }
             }
         }
-
-        if (op == DEL_COMMAND)
+        else if (op == DEL_COMMAND)
         {
             if (exists)
             {
-                removeDecapTunnel(table_name, key);
+                decreaseTunnelRefCount(key);
+                RemoveTunnelIfNotReferenced(key);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Tunnel %s cannot be removed since it doesn't exist.", key.c_str());
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation type %s.", op.c_str());
+        }
+
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+void TunnelDecapOrch::doDecapTunnelTermTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    static const map<string, TunnelTermType> DecapTermTypes = {
+        {"P2P", TUNNEL_TERM_TYPE_P2P},
+        {"P2MP", TUNNEL_TERM_TYPE_P2MP},
+        {"MP2MP", TUNNEL_TERM_TYPE_MP2MP}};
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple &t = it->second;
+
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+
+        string tunnel_name;
+        bool set_src_ip = false;
+        string dst_ip_str;
+        string src_ip_str;
+        IpPrefix dst_ip;
+        IpPrefix src_ip;
+        TunnelTermType term_type = TUNNEL_TERM_TYPE_P2MP;
+        string subnet_type;
+        bool tunnel_exists;
+        bool is_subnet_decap_v4_term;
```

Comment:
Can we use 1 single variable? For example, if `is_subnet_decap_v4_term == true`, then it's subnet_v4, else it's subnet_v6.

Reply:
Updated.